### ISSUE TITLE
Better Github Authorization system

### ISSUE THREAD

**Issue by pragmaticivan** [devspace/devspace-user-requests#34]
Currently the app requires this:
`This application will be able to read and write all public and private repository data. This includes the following:`

I want to be able to choose which repos I want to allow this app to pick data from.

Why should this app have permission to write data in my private repo? 

Why should this app have permission to handle all those things? instead just read them ?

```
Code
Issues
Pull requests
Wikis
Settings
Webhooks and services
Deploy keys
```

**Comment by rimenes** [comment]
Agree on that. Also "read and write all public and **private** repository data" is a no no.

**Comment by zenorocha** [comment]
I agree too! In fact, I had the same initial shock as you all had, that's why I wrote this https://devspace.io/privacy.

In order to show events from private repositories, the app needs to request write access to your public and private repositories. It doesn't make any sense, but that's how GitHub's permission system is, even though the app is 100% readonly :/

**Comment by zenorocha** [comment]
TL;DR GitHub permission system sucks.

In order to show events from private repositories, I need to request write access to your public and private repositories. Even though the app is 100% readonly.

**Comment by pragmaticivan** [comment]
There's no option like that? :
- I just want the app to use my public repo.
- If I want to try private, I can give any permission the app needs.

For me looks like a better approach.

**Comment by zenorocha** [comment]
Yep, that's definitely the best approach.

**Comment by zenorocha** [comment]
As a reference, here are the permission scopes we use: `user` and `repo`.

https://developer.github.com/v3/oauth/#scopes

**Comment by pragmaticivan** [comment]
Cool, I would like to see this feature soon, even if you have a private policy explaining everything, we still don't know how the database, security, access layers of this app infrastructure are now.

For me not having this for this beta version is a little bit dangerous. 

Not familiar with this technique but maybe someone can get in the server, get the token, and simulate requests inside  your app. And with those rights would be pretty easy to download the latest release of a private repo.
(Just speculation haha not familiar with this kind of penetration test)

**Comment by zenorocha** [comment]
There's no such thing. oAuth tokens are generated in the client-side and they're not stored in the server.

**Comment by maracaipe** [comment]
+1 @pragmaticivan.
I don't mind this much with my public repos, but my private. When you offer a way to the app don't need the read/write access to it I would love to beta test this app.

**Comment by zenorocha** [comment]
I talked to GitHub's support but they have no plans to fix this. Working on an open letter now.

**Comment by pragmaticivan** [comment]
@zenorocha you have a simple option for that, two kind of auth. I experienced some apps using a choose button, where you make 2 types of requests, one for just public and one for private. A good example is Waffle.io:

![image](https://cldup.com/iJIm2YAfW--3000x3000.png)

The permission request are different. You can start with that, would be a really awesome feature.

**Comment by zenorocha** [comment]
That's super awesome! Thanks @pragmaticivan

**Comment by shawn42** [comment]
+1 for public only version. Until then I cannot beta test what looks to be a very useful app.

**Comment by zenorocha** [comment]
Totally understand. That's why betas are for, thanks a lot for your feedback guys ;D

**Comment by jfunez** [comment]
Hi @zenorocha this public+private access issue is the only point I've found that don't let me use the app :(
Hope you can fix it soon to test the app.

**Comment by jolaurin** [comment]
+1 for this. I can't wait to test the app. :-)

**Comment by pragmaticivan** [comment]
![forest](https://45.media.tumblr.com/b2d4b1b8de8e60c868f93c52062e0413/tumblr_mg4ex3iRTm1rz1p75o1_400.gif) o/ haha

**Comment by zenorocha** [comment]
Here's a list of other apps that face the same problem :/
- https://github.com/isaacs/github/issues/311
- https://github.com/isaacs/github/issues/31
- https://github.com/huboard/huboard/issues/203
- https://github.com/huboard/huboard/issues/468
- https://github.com/astralapp/astral/issues/11
- https://github.com/bkeepers/github-notifications/issues/146

**Comment by zenorocha** [comment]
Similar login approach on Huboard:

<img width="981" alt="screen shot 2016-02-10 at 9 12 25 pm" src="https://cloud.githubusercontent.com/assets/398893/12969824/10eb4d50-d03b-11e5-82d8-1b0e26432415.png">

**Comment by pragmaticivan** [comment]
Seems to be pretty cool. :+1:

**Comment by zenorocha** [comment]
And here's Gemnasium:

<img width="1000" alt="screen shot 2016-02-10 at 9 20 09 pm" src="https://cloud.githubusercontent.com/assets/398893/12969924/4044e3a8-d03c-11e5-9d77-e9dadb1c7fbe.png">

**Comment by zenorocha** [comment]
Ok, here's the open letter: https://blog.devspace.io/github-s-permission-system-is-flawed-de8836d9e7cc

Join the discussion at: https://github.com/dear-github/dear-github/issues/113

**Comment by damianh** [comment]
I think there will be plenty of organizations that will have difficulty opening up 3rd party access to private repos / data. At this point I can only use the web app for public / oss stuff. The real utility for me would be the private stuff.

For private orgs I'm going to need a desktop app where I configure it with a personal access token that talks directly to the github API. From an org's perspective, this would mean that any security the org applies to specific repos per individual would be respected.

If this should be a different issue, let me know and I'll create.

**Comment by zenorocha** [comment]
Just added a login option for public repos. Thanks all of you for your feedback!

<img width="709" alt="screen shot 2016-02-17 at 11 09 25 am" src="https://cloud.githubusercontent.com/assets/398893/13121392/03d3604a-d567-11e5-81fc-e599ace2f5d2.png">

**Comment by maracaipe** [comment]
Perfect!
